include/picongpu/plugins/PluginRegistry.hpp

```diff
@@ -0,0 +1,198 @@
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "picongpu/particles/traits/SpeciesEligibleForSolver.hpp"
+#include "picongpu/plugins/ISimulationPlugin.hpp"
+
+#include <pmacc/meta/AllCombinations.hpp>
+
+#include <list>
+#include <memory>
+
+
+namespace picongpu
+{
+    using namespace pmacc;
+
+    /** plugin registry entry interface */
+    struct IRegistryEntry
+    {
+        virtual std::shared_ptr<ISimulationPlugin> createPlugin() = 0;
+    };
+
+    /** factory used to create a plugin at runtime */
+    template<typename T_Plugin>
+    struct RegistryEntry : IRegistryEntry
+    {
+        std::shared_ptr<ISimulationPlugin> createPlugin() override
+        {
+            return std::make_shared<T_Plugin>();
+        }
+    };
+
+    /** singleton to register plugins which get created by PluginController during the startup */
+    struct PluginRegistry
+    {
+        /** combines a plugins with all species
+         *
+         * Trait SpeciesEligibleForSolver is evaluated to check if a plugin support using a species.
+         *
+         * @tparam T_Plugin
+         */
+        template<typename T_Plugin>
+        static void registerSpeciesPlugin()
+        {
+            using CombinedUnspecializedSpeciesPlugins = pmacc::AllCombinations<VectorAllSpecies, MakeSeq_t<T_Plugin>>;
+
+            using CombinedUnspecializedSpeciesPluginsEligible
+                = pmacc::mp_copy_if<CombinedUnspecializedSpeciesPlugins, TupleSpeciesPlugin::IsEligible>;
+
+            using SpeciesPlugins
+                = pmacc::mp_transform<TupleSpeciesPlugin::Apply, CombinedUnspecializedSpeciesPluginsEligible>;
+
+
+            meta::ForEach<SpeciesPlugins, PushBack<boost::mpl::_1>> pushBack;
+            pushBack(PluginRegistry::get().pluginList);
+        }
+
+        template<typename T_Plugin>
+        static void registerStandAlonePlugin()
+        {
+            meta::ForEach<MakeSeq_t<T_Plugin>, PushBack<boost::mpl::_1>> pushBack;
+            pushBack(PluginRegistry::get().pluginList);
+        }
+
+        static PluginRegistry& get()
+        {
+            static PluginRegistry instance = PluginRegistry{};
+            return instance;
+        }
+
+        /** get a list of factories to create plugins
+         *
+         * @return list of factories where createPlugin() can be called on each entry
+         */
+        auto getPluginFactories() const
+        {
+            return pluginList;
+        }
+
+    private:
+        std::list<std::shared_ptr<IRegistryEntry>> pluginList;
```

Comment:
Why is this a list and not a vector? And also I think I'd prefer `plugins` over `pluginList` as to [not suggest a type](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rl-name-type). Both are weak non-blocking.

Reply:
It is a list because iteration performance is not important at this point. I can change the type, I copied most of the code from pluginCOntroller where it was a list too.